### PR TITLE
Fix memory leak when there are Joined Arrays in streaming mode

### LINKED ISSUE #3608: Memory leaks in BP5Deserializer::InstallMetaData
This issue is reproducible with latest ADIOS2 2.9.0 and a simple C example of SCORPIO.

Steps to reproduce on a Linux workstation with GCC compiler and its LeakSanitizer:
```
git clone https://github.com/E3SM-Project/scorpio.git
cd scorpio

mkdir build
cd build

CFLAGS="-fsanitize=leak -fno-omit-frame-pointer -g" \
ADIOS2_DIR=/path/to/adios2/installation \
CC=mpicc CXX=mpicxx FC=mpifort cmake -Wno-dev \
-DWITH_ADIOS2=ON \
-DADIOS_BP2NC_TEST=ON \
-DWITH_NETCDF=OFF \
-DPnetCDF_PATH=/path/to/pnetcdf/installation \
-DPIO_USE_MALLOC=ON \
-DPIO_ENABLE_EXAMPLES=ON \
..

make

cd examples/c
mpiexec -n 2 ./example1
```
Example output on a Linux workstation of ANL:
```
=================================================================
==2414827==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 72 byte(s) in 3 object(s) allocated from:
    #0 0x7f80f14a7ac4 in operator new(unsigned long) /nfs/gce/software/tmp/svcbuilder/spack-stage-gcc-11.1.0-qsjmpcgt7go2ourefdh2i5z2g5u7bxmk/spack-src/libsanitizer/lsan/lsan_interceptors.cpp:238
    #1 0x864501 in adios2::format::BP5Deserializer::InstallMetaData(void*, unsigned long, unsigned long, unsigned long) ADIOS2-2.9.0/source/adios2/toolkit/format/bp5/BP5Deserializer.cpp:717
    #2 0x83825e in adios2::core::engine::BP5Reader::InstallMetadataForTimestep(unsigned long) ADIOS2-2.9.0/source/adios2/engine/bp5/BP5Reader.cpp:81
```
Valgrind also reports similar memory leaks:
```
==2420468== 40 (24 direct, 16 indirect) bytes in 1 blocks are definitely lost in loss record 5 of 10
==2420468==    at 0x483BE63: operator new(unsigned long) (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==2420468==    by 0x864501: adios2::format::BP5Deserializer::InstallMetaData(void*, unsigned long, unsigned long, unsigned long) (in scorpio/build/examples/c/example1)
==2420468==    by 0x83825E: adios2::core::engine::BP5Reader::InstallMetadataForTimestep(unsigned long) (in scorpio/build/examples/c/example1)
==2420468==    by 0x83C265: adios2::core::engine::BP5Reader::BeginStep(adios2::StepMode, float) (in scorpio/build/examples/c/example1)
==2420468==    by 0xA055FC: ConvertBPFile(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int) (in scorpio/build/examples/c/example1)
==2420468==    by 0xA073B1: ConvertBPToNC(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int) (in scorpio/build/examples/c/example1)
==2420468==    by 0xA07DBD: C_API_ConvertBPToNC (in scorpio/build/examples/c/example1)
==2420468==    by 0x45D598: PIOc_closefile (in scorpio/build/examples/c/example1)
==2420468==    by 0x45B579: main (in scorpio/build/examples/c/example1)
```
Additional information:
* Reproducible with ADIOS2 2.9.0 (released on Mar 30 2023) and latest master branch
* Not reproducible with ADIOS2 built with master branch on Nov 30 2022
* A similar issue has been fixed for ADIOS2 2.9.0, see #3443

### PR DESCRIPTION
**Pull request by eisenhauer**
Closes #3608

### CODE REVIEW

File: source/adios2/toolkit/format/bp5/BP5Deserializer.h [hunk 1]
```diff
@@ -174,6 +174,7 @@ class BP5Deserializer : virtual public BP5Base
     std::vector<void *> *m_FreeableMBA = nullptr;
 
     std::vector<void *> *m_JoinedDimenOffsetArrays = nullptr;
+    std::vector<void *> *m_FreeableJDOA = nullptr;
```
Comment (vicentebolea) [inline source/adios2/toolkit/format/bp5/BP5Deserializer.h:177]: This attribute needs to the last one in order not to break the ABI compatibility

Comment (vicentebolea) [reply source/adios2/toolkit/format/bp5/BP5Deserializer.h:177]: > would it make sense to put this version in master and merge a version to release_29 with it relocated to the end?

Yes, that is a good idea.

Comment (eisenhauer) [reply source/adios2/toolkit/format/bp5/BP5Deserializer.h:177]: Actually, I think this isn't an ABI issue.  BP5Deserializer is a completely ADIOS-internal class.  You'd only get an ABI issue if this was an externally-facing class and we were changing its public members, right?  Here, if the user ends up pointing to a new version of the ADIOS library he'll end up new code that references this class as well as the new class itself.  Or is there something wrong with my reasoning?

**Comment by pnorbert** [comment]
Can this be merged?

**Comment by vicentebolea** [comment]
> IMHO, yes. @vicentebolea , did some form of this get merged to master? Kind of remember seeing that go by, but I've lost track.

Yes this was added in #3602.

There is a PR fixing this for release_29 too, I am waiting to group together a bunch of bugfixes to backport to the release branch

**Comment by eisenhauer** [comment]
@vicentebolea Just looking back at open PRs.  Can this be closed/deleted now?  I wasn't quite clear based on your May 26 comment, but it sounds like it's in master and that some other PR was putting it in 2.9.

**Comment by vicentebolea** [comment]
@eisenhauer looking deeper into this it appears that what got merged into master/release_29 was a different bugfix, here is: !3619. What do you think?

**Comment by eisenhauer** [comment]
> @eisenhauer looking deeper into this it appears that what got merged into master/release_29 was a different bugfix, here is: #3619. What do you think?

Agreed, that was a different fix.  That fix was for the return-a-structure on compilers that don't support copy elision.  This is a different bug and needs to be in 2.9 and master.  Do you want to batch this in some way?  Or should we update this with current release_29 base and merge?

**Comment by vicentebolea** [comment]
Can you merge to master, I will backported to release.

**Comment by eisenhauer** [comment]
OK, closing this in favor of a new PR targeted at master.

**Pull request closed without merging**